data-prepper-plugins/grok-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepper.java

```diff
@@ -72,18 +79,36 @@ public GrokPrepper(final PluginSetting pluginSetting) {
         final List<Record<String>> recordsOut = new LinkedList<>();
 
         for (final Record<String> record : records) {
+            boolean foundMatch = false;
+
             try {
                 final Map<String, Object> recordMap = OBJECT_MAPPER.readValue(record.getData(), MAP_TYPE_REFERENCE);
+                final Map<String, Object> grokkedLog = new HashMap<>();
 
                 for (final Map.Entry<String, List<Grok>> entry : fieldToGrok.entrySet()) {
                     for (final Grok grok : entry.getValue()) {
                         if (recordMap.containsKey(entry.getKey())) {
                             final Match match = grok.match(recordMap.get(entry.getKey()).toString());
                             match.setKeepEmptyCaptures(grokPrepperConfig.isKeepEmptyCaptures());
 
-                            mergeCaptures(recordMap, match.capture());
+                            final Map<String, Object> captures = match.capture();
+                            mergeCaptures(grokkedLog, captures);
+
+                            if (captures.size() > 0 && grokPrepperConfig.isBreakOnMatch()) {
+                                foundMatch = true;
+                                break;
+                            }
                         }
                     }
+                    if (foundMatch) {
+                        break;
+                    }
```

Comment:
I think that can be improved and remove the variable naming issue for found match by adding a specific function to check if there are captures that were found and the feature is enabled.

```
                            if (shouldBreakOnMatch(grokkedCaptures) {
                                break;
                            }
                        }
                    }
                    if (shouldBreakOnMatch(grokkedCaptures)) {
                        break;
                    }
    ...

private boolean shouldBreakOnMatch(Map<String, Object> captures) {
    return captures.size() > 0 && grokPrepperConfig.isBreakOnMatch();
}